File: frontend/src/api/analytics.ts
/** Analytics API Client.
 *
 * TypeScript interfaces and API functions for fetching analytics metrics
 * from the ARIA analytics endpoints. Covers overview, funnel, trends,
 * response times, ARIA impact, and period comparisons.
 */

import { apiClient } from "./client";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export type AnalyticsPeriod = "7d" | "30d" | "90d" | "all";
export type TrendGranularity = "day" | "week" | "month";

/** High-level overview metrics */
export interface OverviewMetrics {
  leads_created: number;
  meetings_booked: number;
  emails_sent: number;
  debriefs_completed: number;
  goals_completed: number;
  avg_health_score: number | null;
  time_saved_minutes: number;
}

/** Conversion funnel stage data */
export interface ConversionFunnel {
  stages: Record<string, number>;
  conversion_rates: Record<string, number | null>;
  avg_days_in_stage: Record<string, number | null>;
}

/** Activity trends time series */
export interface ActivityTrends {
  granularity: TrendGranularity;
  series: {
    emails_sent: Record<string, number>;
    meetings: Record<string, number>;
    aria_actions: Record<string, number>;
    leads_created: Record<string, number>;
  };
}

/** Response time metrics */
export interface ResponseTimeMetrics {
  avg_response_minutes: number | null;
  by_lead: Record<string, number>;
  trend: Array<{ date: string; avg_response_minutes: number }>;
}

/** ARIA impact summary */
export interface AriaImpactSummary {
  total_actions: number;
  by_action_type: Record<string, number>;
  estimated_time_saved_minutes: number;
  pipeline_impact: Record<string, { count: number; estimated_value: number }>;
}

/** Period comparison with deltas */
export interface PeriodComparison {
  current: OverviewMetrics;
  previous: OverviewMetrics;
  delta_pct: Record<string, number | null>;
}

// ---------------------------------------------------------------------------
// API Functions
// ---------------------------------------------------------------------------

/** Get high-level overview metrics */
export async function getOverviewMetrics(
  period: AnalyticsPeriod = "30d",
): Promise<OverviewMetrics> {
  const response = await apiClient.get<OverviewMetrics>("/analytics/overview", {
    params: { period },
  });
  return response.data;
}

/** Get conversion funnel metrics */
export async function getConversionFunnel(
  period: AnalyticsPeriod = "30d",
): Promise<ConversionFunnel> {
  const response = await apiClient.get<ConversionFunnel>("/analytics/funnel", {
    params: { period },
  });
  return response.data;
}

/** Get activity trends */
export async function getActivityTrends(
  period: AnalyticsPeriod = "30d",
  granularity: TrendGranularity = "day",
): Promise<ActivityTrends> {
  const response = await apiClient.get<ActivityTrends>("/analytics/trends", {
    params: { period, granularity },
  });
  return response.data;
}

/** Get response time metrics */
export async function getResponseTimeMetrics(
  period: AnalyticsPeriod = "30d",
): Promise<ResponseTimeMetrics> {
  const response = await apiClient.get<ResponseTimeMetrics>(
    "/analytics/response-times",
    {
      params: { period },
    },
  );
  return response.data;
}

/** Get ARIA impact summary */
export async function getAriaImpactSummary(
  period: AnalyticsPeriod = "30d",
): Promise<AriaImpactSummary> {
  const response = await apiClient.get<AriaImpactSummary>(
    "/analytics/aria-impact",
    {
      params: { period },
    },
  );
  return response.data;
}

/** Compare metrics between two periods */
export async function comparePeriods(
  currentPeriod: AnalyticsPeriod = "30d",
  previousPeriod: AnalyticsPeriod = "30d",
): Promise<PeriodComparison> {
  const response = await apiClient.get<PeriodComparison>("/analytics/compare", {
    params: { current: currentPeriod, previous: previousPeriod },
  });
  return response.data;
}

/** Export analytics data as CSV or JSON */
export async function exportAnalytics(
  period: AnalyticsPeriod = "30d",
  format: "csv" | "json" = "csv",
): Promise<void> {
  const response = await apiClient.get("/analytics/export", {
    params: { period, format },
    responseType: "blob",
  });

  const blob = new Blob([response.data], {
    type: format === "csv" ? "text/csv" : "application/json",
  });
  const url = window.URL.createObjectURL(blob);
  const link = document.createElement("a");
  link.href = url;
  link.setAttribute(
    "download",
    `aria-analytics-${period}.${format}`,
  );
  document.body.appendChild(link);
  link.click();
  document.body.removeChild(link);
  window.URL.revokeObjectURL(url);
}

File: backend/src/models/analytics.py
"""Pydantic models for Analytics API responses."""

from pydantic import BaseModel, Field


class OverviewMetricsResponse(BaseModel):
    """High-level overview metrics response."""

    leads_created: int = Field(..., ge=0, description="Number of leads created in period")
    meetings_booked: int = Field(
        ..., ge=0, description="Number of meetings booked with external attendees"
    )
    emails_sent: int = Field(..., ge=0, description="Number of emails sent in period")
    debriefs_completed: int = Field(..., ge=0, description="Number of meeting debriefs completed")
    goals_completed: int = Field(..., ge=0, description="Number of goals marked complete")
    avg_health_score: float | None = Field(
        None, ge=0, le=100, description="Average health score across active leads"
    )
    time_saved_minutes: int = Field(
        ..., ge=0, description="Total estimated minutes saved from ARIA actions"
    )


class ConversionFunnelResponse(BaseModel):
    """Conversion funnel metrics response."""

    stages: dict[str, int] = Field(
        default_factory=dict,
        description="Count of leads in each lifecycle stage (lead, opportunity, account)",
    )
    conversion_rates: dict[str, float | None] = Field(
        default_factory=dict,
        description="Stage-to-stage conversion rates (e.g., lead_to_opportunity)",
    )
    avg_days_in_stage: dict[str, float | None] = Field(
        default_factory=dict,
        description="Average days spent in each stage",
    )


class ActivityTrendPoint(BaseModel):
    """Single data point in activity trend series."""

    date: str = Field(
        ..., description="Date bucket (YYYY-MM-DD, YYYY-MM-DD for week, or YYYY-MM for month)"
    )
    count: int = Field(..., ge=0, description="Count of activities in this bucket")


class ActivityTrendsResponse(BaseModel):
    """Activity trends time series response."""

    granularity: str = Field(..., description="Time grouping: 'day', 'week', or 'month'")
    series: dict[str, dict[str, int]] = Field(
        default_factory=dict,
        description="Time series for each metric (emails_sent, meetings, aria_actions, leads_created)",
    )


class ResponseTimeTrendPoint(BaseModel):
    """Single data point in response time trend."""

    date: str = Field(..., description="Date (YYYY-MM-DD)")
    avg_response_minutes: float = Field(..., ge=0, description="Average response time in minutes")


class ResponseTimeMetricsResponse(BaseModel):
    """Response time metrics response."""

    avg_response_minutes: float | None = Field(
        None, ge=0, description="Overall average response time in minutes"
    )
    by_lead: dict[str, float] = Field(
        default_factory=dict,
        description="Average response time per lead (lead_memory_id -> minutes)",
    )
    trend: list[ResponseTimeTrendPoint] = Field(
        default_factory=list,
        description="Daily response time trend",
    )


class PipelineImpactBreakdown(BaseModel):
    """Pipeline impact breakdown for a single impact type."""

    count: int = Field(..., ge=0, description="Number of impacts of this type")
    estimated_value: float = Field(..., ge=0, description="Total estimated value in dollars")


class AriaImpactSummaryResponse(BaseModel):
    """ARIA impact summary response."""

    total_actions: int = Field(..., ge=0, description="Total ARIA actions taken")
    by_action_type: dict[str, int] = Field(
        default_factory=dict,
        description="Breakdown of actions by type (email_draft, meeting_prep, etc.)",
    )
    estimated_time_saved_minutes: int = Field(
        ..., ge=0, description="Total estimated minutes saved"
    )
    pipeline_impact: dict[str, PipelineImpactBreakdown] = Field(
        default_factory=dict,
        description="Pipeline impact breakdown by type",
    )


class PeriodComparisonResponse(BaseModel):
    """Period-over-period comparison response."""

    current: OverviewMetricsResponse = Field(..., description="Metrics for current period")
    previous: OverviewMetricsResponse = Field(..., description="Metrics for previous period")
    delta_pct: dict[str, float | None] = Field(
        default_factory=dict,
        description="Percentage change for each metric (absolute for avg_health_score)",
    )
